validate/src/command.rs

```diff
@@ -250,22 +279,70 @@ pub fn name(value: impl AsRef<str>) -> Result<(), CommandValidationError> {
         Ok(())
     } else {
         Err(CommandValidationError {
-            kind: CommandValidationErrorType::NameInvalid,
+            kind: CommandValidationErrorType::NameLengthInvalid,
         })
     }
 }
 
+/// Validate the name of a [`ChatInput`] command or a [`CommandOption`] name.
+///
+/// The length of the name must be more than [`NAME_LENGTH_MIN`] and less than
+/// or equal to [`NAME_LENGTH_MAX`]. It can only contain alphanumeric characters
+/// and lowercase variants must be used where possible. Special characters `-`
+/// and `_` are allowed.
+///
+/// # Errors
+///
+/// Returns an error of type [`NameLengthInvalid`] if the length is invalid.
+///
+/// Returns an error of type [`NameCharacterInvalid`] if the name contains
+/// a non-alphanumeric character or an uppercase character for which a
+/// lowercase variant exists.
+///
+/// [`ChatInput`]: CommandType::ChatInput
+/// [`NameLengthInvalid`]: CommandValidationErrorType::NameLengthInvalid
+/// [`NameCharacterInvalid`]: CommandValidationErrorType::NameCharacterInvalid
+pub fn chat_input_name(value: impl AsRef<str>) -> Result<(), CommandValidationError> {
+    let len = value.as_ref().chars().count();
+
+    if !(NAME_LENGTH_MIN..=NAME_LENGTH_MAX).contains(&len) {
+        return Err(CommandValidationError {
+            kind: CommandValidationErrorType::NameLengthInvalid,
+        });
+    }
+
+    let chars = value.as_ref().chars();
+
+    for char in chars {
+        if !char.is_alphanumeric() && char != '_' && char != '-' {
+            return Err(CommandValidationError {
+                kind: CommandValidationErrorType::NameCharacterInvalid { character: char },
+            });
+        }
+
+        if char.to_lowercase().next() != Some(char) {
```

Comment:
should we just use [`char::is_lowercase`] here?

[`char::is_lowercase`]: https://doc.rust-lang.org/stable/std/primitive.char.html#method.is_lowercase

Reply:
Some characters do not have case (such as some Chinese characters), the actual condition fail only if a character has a lowercase variant.